parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java

```diff
@@ -116,7 +116,18 @@ private boolean eval(MessageType fileSchema, BlockMetaData rowGroup,
         }
       }
 
-      return ExpressionVisitors.visitEvaluator(expr, this);
+      try {
+        return ExpressionVisitors.visitEvaluator(expr, this);
+
+      } finally {
+        // allow temporary state to be collected because this is in a thread-local
```

Comment:
As it resets all instance variables, curious what if directly creating a new `EvalVisitor` object instead of using ThreadLocal.

Reply:
Good question. I think we could remove the thread-local. The idea was originally to avoid the `Expression` handling that is currently done in the constructor: `Binder.bind(struct, Expressions.rewriteNot(unbound), caseSensitive)`.

We could remove the thread-local since it is just avoiding a simple object allocation that's probably not worth it. On the other hand, we could also keep it and simply clear all of this state instead of allocating new containers each time. I'd probably opt for that to avoid lots of object creation. What do you think?

Reply:
I think it makes sense to fix it for now with the current way, which has the minimal side effect.

Additionally, I think we might not save much in case there is a new thread for each task. Also the thread local itself might add additional cost as well.